Table Block: Fixed width table cells on by default

t-hamano opened the pull request:
Fixes: #16045
Alternative to #18870

## What?
This PR turns on fixed table layout by default in the Table block.

## Why?

I believe there is already consensus on this change, but I personally also think the default behavior of changing cell widths as text is entered is a bit confusing.

## How?
With the addition of the deprecation, I have updated the fixture files, unit tests and e2e test snapshots.

### Fixtures

Previous blocks with a default value of `false` for `hasFixedLayout` will have `{hasFixedLayout: false}` added as a comment delimiter by passing this deprecation. Therefore, I have added this comment delimiter to the existing fixture htmls.

For the v4 deprecation I added this time, I added a test scenario as `core__table__deprecated-4.html`.

### Unit Tests

The table block has two types of unit tests to ensure that data from the document is correctly pasted into the table:

- [test/integration/blocks-raw-handling.test.js](https://github.com/WordPress/gutenberg/blob/7810f7bdc29f3ba798aee28e5c70c97d8825fda0/test/integration/blocks-raw-handling.test.js#L380-L383)
- [packages/blocks/src/api/raw-handling/test/paste-handler.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/api/raw-handling/test/paste-handler.js)

I have updated these files so that the default value of `hasFixedLayout` is `true`.

### E2E Tests

Added `has-fixed-layout` class to the table element in the snapshot.

t-hamano committed:
Table Block: Fixed width table cells on by default

aaronrobertshaw commented on test/integration/fixtures/blocks/core__table.html:
```diff
@@ -1,3 +1,3 @@
-<!-- wp:core/table -->
+<!-- wp:core/table {"hasFixedLayout":false} -->
```
These fixtures should reflect the current default state unless they target a specific use case. I think the better approach is to update the fixture e.g. markup etc.

aaronrobertshaw commented on test/integration/fixtures/blocks/core__table__deprecated-1.html:
```diff
@@ -1,3 +1,3 @@
-<!-- wp:table {"backgroundColor":"subtle-pale-blue"} -->
+<!-- wp:table {"hasFixedLayout":false,"backgroundColor":"subtle-pale-blue"} -->
```
This change means the fixture no longer represents as accurately what the block attributes were at the time it was the current version. 

In this case, the deprecated version relied on the default for `hasFixedLayout` being `false.`. As each deprecation should bring the block up to the latest version including changing the default for `hasFixedLayout`, leaving this as it was should ensure the deprecation is working correctly.

aaronrobertshaw commented on packages/block-library/src/table/deprecated.js:
```diff
@@ -24,6 +25,297 @@ const oldColors = {
+			default: [],
+			source: 'query',
+			selector: 'thead tr',
+			query: {
```
Minor suggestion, feel free to ignore this:

Each of these query objects are quite verbose. Given the deprecation isn't generally a point of reference is it worth extracting this to make things a little more concise and easier to scan?

t-hamano replied:
In v2/v4, this `query` object has changed. Therefore, I think it may be difficult to extract this query.

aaronrobertshaw replied:
Sorry, I could have been clearer. 

I didn't mean all query objects had to be a single variable. Each deprecation repeats the same query three times, for the `head`, `body`, and `foot` attributes.

t-hamano replied:
I used `reduce()`to combine `head`, `body`, and `foot` into one. You can check the diff [here](https://github.com/WordPress/gutenberg/pull/49672/commits/679c662dd5d36e96824bb6a6bede8d57a0158381). Is this approach correct?

aaronrobertshaw replied:
Sorry, this minor suggestion seems to have derailed things here 🙏 

The goal was to make the deprecated code simple to grok, reducing cognitive load, by making it a little more concise and less repetitive. A reason is, that having to scroll repeatedly to check if the query properties in the attributes are the same is tedious and error-prone.

The use of `reduce`, while being the most concise and elegant result possible, makes grokking the attributes more difficult for most people. Deprecations are difficult enough without having to devote too much brain power to just understanding the attributes 😅 

<details>
<summary>Example of the compromise I had in mind</summary>

```js
const v4Query = {
	content: {
		type: 'rich-text',
		source: 'rich-text',
	},
	tag: {
		type: 'string',
		default: 'td',
		source: 'tag',
	},
	scope: {
		type: 'string',
		source: 'attribute',
		attribute: 'scope',
	},
	align: {
		type: 'string',
		source: 'attribute',
		attribute: 'data-align',
	},
	colspan: {
		type: 'string',
		source: 'attribute',
		attribute: 'colspan',
	},
	rowspan: {
		type: 'string',
		source: 'attribute',
		attribute: 'rowspan',
	},
};

const v4 = {
	attributes: {
		...
		head: {
			type: 'array',
			default: [],
			source: 'query',
			selector: 'thead tr',
			query: {
				cells: {
					type: 'array',
					default: [],
					source: 'query',
					selector: 'td,th',
					query: v4Query,
				},
			},
		},
		...
	},
};
```
</details>

Perhaps this going off the rails a little bit shows we should just leave the attributes alone then? At the least, I'd vote for reverting the `reduce` refactor.

What do you think?

t-hamano replied:
Thank you for the detailed explanation! I reverted the `reduce` refactor and tried incorporating your approach in [d201b1c](https://github.com/WordPress/gutenberg/pull/49672/commits/d201b1cb534d8c82dd4269ead175f771d9c251aa).

aaronrobertshaw reviewed (commented):
Thanks for putting this together @t-hamano 👍 

It's coming together well so far but there are a couple of tweaks I think we need to make around the deprecation and fixtures.

#### Deprecation
✅ New deprecation attributes match correctly
✅ Save function matches
❓ Deprecation `supports` are missing the `interactivity` support

#### Fixtures
❓ The non-deprecation fixtures should reflect "default" attributes other than the use case they are target i.e. they shouldn't add `hasFixedLayout: false` but instead have their markup etc updated to reflect the latest.
❓ For the deprecation fixtures they should remain with the defaults they had at the time that block was current and migrate all the way to the latest i.e. they shouldn't have their `hasFixedLayout` set either.
❓ We should probably also be consistent between the block fixtures and the document fixtures i.e. rely on the default attribute as they both did originally.

t-hamano committed:
Merge branch 'trunk' into update/table-default-fixed-cell

t-hamano committed:
Add interactivity support to deprecation

t-hamano committed:
Revert some fixture changes

t-hamano committed:
Update some fixtures

t-hamano commented:
@aaronrobertshaw Thanks for the review!

I made additional commits for:

- Added `interactivity` to the deprecation `support`.
- I removed the `hasFixedLayout` comment attribute in the fixture file below.
  - `core__table__deprecated-1.html`
  - `core__table__deprecated-2.html`
  - `core__table__deprecated-3.html`
- I removed the `hasFixedLayout` comment attribute and updated the markup (added the `has-fixed-layout` class to the table element) in the fixture file below:
  - `core__table.html`
  - `core__table__caption.html`
  - `core__table__scope-attribute.html`

If I don't understand correctly, please let me know 🙏

aaronrobertshaw reviewed (approved):
Thanks for iterating on this @t-hamano 👍 

> If I don't understand correctly, please let me know

It looks like all the major stuff has been sorted. I think this is pretty much ready to go.

✅ New deprecation's supports match (interactivity support has been added)
✅ Default (current version) fixtures are correct
✅ New deprecation is covered by fixtures
✅ Consistency between snapshots and fixtures
✅ Fixture tests are passing
✅ Manually tests as advertised

<img width="816" alt="Screenshot 2024-03-20 at 11 25 21 AM" src="https://github.com/WordPress/gutenberg/assets/60436221/1a2a472b-180a-4f16-b169-33157f5d4fbb">

LGTM 🚢

t-hamano committed:
Simplify attributes

t-hamano committed:
Revert "Simplify attributes"

This reverts commit 679c662dd5d36e96824bb6a6bede8d57a0158381.

t-hamano committed:
Merge branch 'trunk' into update/table-default-fixed-cell

t-hamano committed:
Extract cell query

aaronrobertshaw commented:
@t-hamano I've given this another run after the latest changes. LGTM 🚢
